system/jlib/jevent.cpp

```diff
@@ -930,75 +930,621 @@ bool readEvents(const char* filename, IEventVisitor& visitor)
+    }
+
+    virtual void recordAttribute(const char* name, const char* value, bool quoted) = 0;
+};
+
+// Concrete extension of TDumpEventVisitor<IPTreeEventVisitor> that uses an IPropertyTree to store
+// visited data. The tree is constructed in memory and is available via the `queryTree` method
+// after visitation.
+class CPTreeEventVisitor : public TDumpEventVisitor<IPTreeEventVisitor>
+{
+public: // IEventVisitor
     virtual bool visitFile(const char* filename, uint32_t version) override
     {
-        *out << "name: " << filename << std::endl;
-        *out << "version: " << version << std::endl;
+        tree.setown(createPTree(DUMP_STRUCTURE_ROOT));
+        active = tree->addPropTree(DUMP_STRUCTURE_HEADER, createPTree());
+        doVisitHeader(filename, version);
         return true;
     }
+
     virtual Continuation visitEvent(EventType id) override
     {
-        *out << "event: " << queryEventName(id) << std::endl;
+        active = tree->addPropTree(DUMP_STRUCTURE_EVENT, createPTree());
+        doVisitEvent(id);
         return visitContinue;
     }
-    virtual Continuation visitAttribute(EventAttr id, const char * value) override
+
+    virtual bool departEvent() override
     {
-        *out << "attribute: " << queryEventAttributeName(id) << " = '" << value << "'" << std::endl;
-        return visitContinue;
+        active = tree.get();
+        return true;
     }
-    virtual Continuation visitAttribute(EventAttr id, bool value) override
+
+    virtual void departFile(uint32_t bytesRead) override
     {
-        *out << "attribute: " << queryEventAttributeName(id) << " = " << (value ? "true" : "false") << std::endl;
-        return visitContinue;
+        active = tree->addPropTree(DUMP_STRUCTURE_FOOTER, createPTree());
+        doVisitFooter(bytesRead);
     }
-    virtual Continuation visitAttribute(EventAttr id, uint8_t value) override
+
+    virtual IPTree* queryTree() const override
     {
-        *out << "attribute: " << queryEventAttributeName(id) << " = " << (unsigned)value << std::endl;
-        return visitContinue;
+        return tree.get();
     }
-    virtual Continuation visitAttribute(EventAttr id, uint16_t value) override
+
+protected:
+    virtual void recordAttribute(const char* name, const char* value, bool quoted) override
     {
-        *out << "attribute: " << queryEventAttributeName(id) << " = " << (unsigned)value << std::endl;
+        active->addProp(VStringBuffer("@%s", name), value);
+    }
+
+protected:
+    Owned<IPTree> tree;
+    IPTree* active = nullptr;
+};
+
+IPTreeEventVisitor* createPTreeEventVisitor()
+{
+    return new CPTreeEventVisitor();
+}
+
+// Concrete extension of TDumpEventVisitor<> that writes visited data to an output stream. One line
+// of text is produced for each event and attribute visited, plus additional lines for the file
+// name, version, and number of bytes read from the file.
+class CDumpTextEventVisitor : public TDumpEventVisitor<>
```

Comment:
This class seems to be a mixture of a base class and a final class.  If so it would be clearer if they were separate.

Reply:
It is both. I wasn't sure that a base with just the dump method and the shared data would be better.